scalafix-core/shared/src/main/scala/scalafix/internal/patch/EscapeHatch.scala

```diff
@@ -0,0 +1,97 @@
+package scalafix
+package internal.patch
+
+import scala.meta._
+
+object RangePosition {
+  def apply(position: Position): RangePosition =
+    RangePosition(position.start, position.end)
+}
+
+case class RangePosition(start: Int, end: Int) {
+  def withIn(other: RangePosition): Boolean = {
+    start <= other.start && other.end <= end
+  }
+}
+case class PointPosition(offset: Int)
+
+case class CommentFilter(
+    start: PointPosition,
+    end: Option[PointPosition],
+    rules: Set[String]
+) {
+  def isEnabled(ruleName: String, rulePosition: RangePosition): Boolean = {
+    val ruleMatches =
+      if (rules.nonEmpty) rules.contains(ruleName)
+      else true
+
+    ruleMatches && intersect(rulePosition)
+  }
+
+  private def intersect(rulePosition: RangePosition): Boolean = {
+    end match {
+      case Some(endPos) => {
+        val range = RangePosition(start.offset, endPos.offset)
+        range.withIn(rulePosition)
+      }
+      case None => {
+        start.offset <= rulePosition.start
+      }
+    }
+  }
+}
+
+sealed trait FilterAnchor {
+  def position: PointPosition
+}
+
+object FilterAnchor {
+  case class Start(position: PointPosition, rules: Set[String])
+      extends FilterAnchor
+  case class End(position: PointPosition) extends FilterAnchor
+}
+
+object EscapeHatch {
+  private val FilterAnchorStart = "scalafix:off (.*)".r
+  private val FilterAnchorEnd = "scalafix:on".r
+
+  def apply(tree: Tree): EscapeHatch = {
+    val anchors: List[FilterAnchor] =
+      tree.tokens.collect {
+        case comment @ Token.Comment(FilterAnchorStart(rules)) => {
+          FilterAnchor.Start(
+            PointPosition(comment.pos.end),
+            rules.split("\\s+").toSet
+          )
+        }
+
+        case comment @ Token.Comment(FilterAnchorEnd()) => {
+          FilterAnchor.End(
+            PointPosition(comment.pos.start)
+          )
+        }
+      }.toList
+
+    val filters: List[CommentFilter] =
+      anchors.sortBy(_.position.offset).grouped(2).toList.collect {
+        case List(FilterAnchor.Start(start, rules), FilterAnchor.End(end)) => {
+          CommentFilter(start, Some(end), rules)
+        }
+
+        case List(FilterAnchor.Start(start, rules)) => {
+          CommentFilter(start, None, rules)
+        }
+
+        case _ => sys.error("invalid anchors")
```

Comment:
Can we include the anchor in the message to help debug what went wrong?

Reply:
We are inside a collect, why do we fail here?

Reply:
What happens in this case?
```scala
// scalafix:off A
// scalafix:off B
// scalafix:on B
// scalafix:on A
```

Reply:
Done. Added test case.